src/main/kotlin/com/pubnub/api/retry/RetryableBase.kt

```diff
@@ -0,0 +1,111 @@
+            500 to "HTTP_INTERNAL_ERROR",
+            501 to "HTTP_NOT_IMPLEMENTED",
+            502 to "HTTP_BAD_GATEWAY",
+            503 to "HTTP_UNAVAILABLE",
+            504 to "HTTP_GATEWAY_TIMEOUT",
+            505 to "HTTP_VERSION",
+        )
+        internal val retryableExceptions = listOf(
+            UnknownHostException::class.java,
+            SocketTimeoutException::class.java,
+            ConnectException::class.java,
+            SSLHandshakeException::class.java,
+        )
+    }
+
+    private var exponentialMultiplier = 0.0
+
+    internal fun isRetryPolicySetForThisRestCall(): Boolean {
+        return when (retryPolicy) {
+            is RequestRetryPolicy.None -> false
+
+            is RequestRetryPolicy.Linear -> {
+                val excludedOperations = retryPolicy.excludedOperations
+                endpointIsNotExcludedFromRetryPolicy(excludedOperations)
+            }
+            is RequestRetryPolicy.Exponential -> {
+                val excludedOperations = retryPolicy.excludedOperations
+                endpointIsNotExcludedFromRetryPolicy(excludedOperations)
+            }
+        }
+    }
+
+    internal fun getDelayInMilliSeconds(response: Response<T>): Int {
+        val effectiveDelay = if (response.raw().code == TOO_MANY_REQUEST) {
+            calculateDelayForTooManyRequestError(response)
+        } else {
+            val delayFromRetryPolicyInMilliSec = getDelayFromRetryPolicy() * MILLISECONDS
+            delayFromRetryPolicyInMilliSec
+        }
+        return effectiveDelay
+    }
+
+    internal fun getDelayFromRetryPolicy(): Int {
+        return when (retryPolicy) {
+            is RequestRetryPolicy.None -> 0
+            is RequestRetryPolicy.Linear -> retryPolicy.delayInSec
+            is RequestRetryPolicy.Exponential -> {
+                val delay: Int = (retryPolicy.minDelayInSec * 2.0.pow(exponentialMultiplier)).toInt()
+                exponentialMultiplier++
+                minOf(delay, retryPolicy.maxDelayInSec)
+            }
+        }
+    }
+
+    internal fun getRetryCount(): Int {
+        return when (retryPolicy) {
+            is RequestRetryPolicy.None -> 0
+            is RequestRetryPolicy.Linear -> retryPolicy.maxRetryNumber
+            is RequestRetryPolicy.Exponential -> retryPolicy.maxRetryNumber
+        }
+    }
+
+    internal fun isErrorCodeRetryable(errorCode: Int) = retryableStatusCodes.containsKey(errorCode)
+
+    private fun calculateDelayForTooManyRequestError(
+        response: Response<T>
+    ): Int {
+        val retryAfterInSec: String? = response.headers()[RETRY_AFTER_HEADER_NAME]
+        if (!retryAfterInSec.isNullOrBlank() && !isNumeric(retryAfterInSec)) {
+            throw PubNubException(PubNubError.RETRY_AFTER_HEADER_VALUE_CAN_NOT_BE_PARSED_TO_INT)
+        }
+        return retryAfterInSec?.toIntOrNull()?.let { retryAfterInSec ->
+            if (retryAfterInSec > 0) {
+                retryAfterInSec * MILLISECONDS
```

Comment:
I am not sure. So you are proposing to change from:
getDelayFromRetryPolicy() * MILLISECONDS
to 
getDelayFromRetryPolicy().seconds.inWholeMilliseconds.toInt()
?
What is an advantage of this approach? 
For me it doesn't look better.

Reply:
yeah you're right it does look long in this case... what I don't like is creating a new MILLISECONDS constant in files whenever we need to convert time units, when we have code that does exactly that in kotlin.

the idea would be to not use int or long until the very last moment you need it, and improve type-safety instead of passing around various ints and longs. See your variable name for example: retryAfter**inSec**. You wouldn't need that, if it was just a Duration. 
Functions that return delays, durations etc. could return a Duration (so you'd call .seconds) INSIDE getDelayFromRetryPolicy and you operate on Durations (which prevents you from making an error and mixing seconds with milliseconds on accident) until you need an int/long, such as for Thread.sleep().

Reply:
I like this proposal. I will change the code.